### ISSUE TITLE
What are the ways of treatng missing values in XGboost?

### ISSUE THREAD

**Issue by naggar1** [dmlc/xgboost#21]
Generally does the model performance get better with that ?

**Comment by gitowes** [comment]
Looks like in the tree code, there is logic handling unknowns. However, in the io code, I haven't spotted where the missing values are set. It seems to me that all missing values will be treated as zeros by the resize functions. Is that correct?

**Comment by Acriche** [comment]
A follow up question-
While I understand how XGboost handles missing values within discrete variables, I'm not sure how does it handle continues (numeric) variables.
Can you please explain?

**Comment by tqchen** [comment]
For continuous features, a missing(default) direction is learnt for missing value data to go into, so when the data of the speficific value is missing, then it goes to the default direction

**Comment by Acriche** [comment]
Thanks Tianqi.
And what about missing continuous features in generalized linear models?

**Comment by akshenndra** [comment]
Hi Tianqi
I am looking for an algo which does no imputation of the missing values internally and yet works .How does the xgboost work internally to handle missing values(can you drop in some basic idea) ?

**Comment by tqchen** [comment]
see https://arxiv.org/abs/1603.02754 sec 3.4

**Comment by akshenndra** [comment]
Xgboost also works in the presence of categorical features ?We don't need to prepocess them (binarisation,etc.).For e.g my dataset has a feature called city which has values-"Milan","Rome","venice".Can I present them to xgboost without any preproccessing at all?

**Comment by johnsonr05** [comment]
Tianqi,

I have question about xgb.importance function. When I run this and look at the Real Cover it seems as though if there is any missing data in a feature, the Real Cover is NA. Is there anyway to deal with this issue to get some co-occurence count for each split?

Rex

**Comment by nyutal** [comment]
Hi Tianqi, 
my processing pipeline include normalize the features before learning. Also, i have a lot of indicators which are missing and not zero for negative indication. 
as a result, my normalized indicators are 0 for indicated values and missing for non-indicated values.
will xgboost handle such behavior properly? (changing non-exists features to 0 will cause problems...)

Thanks,
Nadav

**Comment by Damdinov** [comment]
@tqchen 
You said
> Internally, XGBoost will automatically learn what is the best direction to go when a value is missing. Equivalently, this can be viewed as automatically "learn" what is the best imputation value for missing values based on reduction on training loss.

What about a case when the train set has not missing values, but the test has?
